ssl/quic/quic_port.c

```diff
@@ -1336,6 +1357,183 @@ static int port_validate_token(QUIC_PKT_HDR *hdr, QUIC_PORT *port,
+ */
+static int port_chk_initial_pkt(QUIC_PORT *port, QUIC_URXE *e,
+                                QUIC_PKT_HDR *hdr, QUIC_PORT_SECRETS **pqps)
+{
+    PACKET pkt;
+    QUIC_PKT_HDR  hdr_local = { 0 };
+    QUIC_PKT_HDR_PTRS ptrs = { 0 };
+    unsigned char *work_buf = NULL;
+    size_t work_buf_sz = 0;
+    unsigned char scratch_dst[65535];
+    unsigned char nonce[EVP_MAX_IV_LENGTH];
+    const unsigned char *sop;
+    QUIC_PN pn;
+    size_t aad_len;
+    int l = 0, l2 = 0;
+    int ok;
+    QUIC_PORT_SECRETS *qps = NULL;
+
+    if (hdr->type != QUIC_PKT_TYPE_INITIAL)
+        return 0;
+
+    if (!PACKET_buf_init(&pkt, ossl_quic_urxe_data(e), e->data_len))
+        return 0;
+
+    if (pqps == NULL)
+        return 0;
+
+    /*
+     * create packet copy so qrx will obtain raw packet to do its own
+     * processing. This is sub-optimal. Ideally we want port to create qrx
+     * instance here to process packet.  if port will accept packet and create
+     * a channel later, then the qrx instance will be passed to channel
+     * constructor.
+     */
+    if (PACKET_memdup(&pkt, &work_buf, &work_buf_sz) == 0)
+        return 0;
+
+    if (!PACKET_buf_init(&pkt, work_buf, work_buf_sz))
+        goto fail;
+
+    hdr = &hdr_local;
+    if (ossl_quic_wire_decode_pkt_hdr(&pkt, SIZE_MAX, 1, 0, hdr, &ptrs,
+                                      NULL) == 0)
+        goto fail;
+
+    if (ptrs.raw_sample_len < 16)
+        goto fail;
+
+    /*
+     * code here got copied and modified from
+     * ossl_quic_hdr_protector_decrypt_fields()
+     */
+    qps = ossl_quic_provide_initial_rx_secret_for_dcid(port, &hdr->dst_conn_id);
+    if (qps == NULL)
+        goto fail;
+
+    if (port_chk_remove_hdr_protection(qps, &ptrs) == 0)
+        goto fail;
+
+    /* decode header with protection removed  */
+    if (!PACKET_buf_init(&pkt, work_buf, work_buf_sz)) {
+        ossl_quic_destroy_port_secrets(qps);
+        goto fail;
+    }
+    sop = PACKET_data(&pkt);
+    ok = ossl_quic_wire_decode_pkt_hdr(&pkt, SIZE_MAX, 0, 0, hdr, NULL, NULL);
+    if (ok == 0)
+        goto fail;
+
+    ok = ossl_quic_wire_decode_pkt_hdr_pn(hdr->pn, hdr->pn_len, 0, &pn);
+    if (ok == 0)
+        goto fail;
+
+    /*
+     * code here got copied and modified from qrx_decrypt_pkt_body()
+     * and el_setup_keyslot(). it supports initial level encryption only
+     * for rx only packets.
+     */
+
+    if (port_chk_derive_nonce(qps, pn, nonce, EVP_MAX_IV_LENGTH) == 0)
+        goto fail;
+
+    /* type and key  got up in ossl_quic_provide_initial_rx_secret_for_dcid. */
+    ok = EVP_CipherInit_ex(qps->qps_pkt_ctx, NULL,
+                           NULL, NULL, nonce, /*enc=*/0);
+    if (ok == 0)
+        goto fail;
+
+    /* Feed the AEAD tag we got so the cipher can validate it. */
+    ok = EVP_CIPHER_CTX_ctrl(qps->qps_pkt_ctx, EVP_CTRL_AEAD_SET_TAG, 16,
+                             (unsigned char *)hdr->data + hdr->len - 16);
```

Comment:
is 16 bytes always the size of every aead ciphers tag length?  Should we be using EVP_CIPHER_CTX_get_tag_length here to get the appropriate tag size?